libcloud/compute/drivers/outscale.py

```diff
@@ -0,0 +1,1018 @@
+
+import json
+
+import requests
+
+from libcloud.compute.base import NodeDriver
+from libcloud.compute.types import Provider
+from libcloud.common.osc import OSCRequestSignerAlgorithmV4
+from libcloud.common.base import ConnectionUserAndKey
+from libcloud.compute.base import \
+    Node,\
+    NodeImage, \
+    KeyPair, \
+    StorageVolume, \
+    VolumeSnapshot, \
+    NodeLocation
+from libcloud.compute.types import NodeState
+
+
+class OutscaleNodeDriver(NodeDriver):
+    """
+    Outscale SDK node driver
+    """
+
+    type = Provider.OUTSCALE
+    name = 'Outscale API'
+    website = 'http://www.outscale.com'
+
+    def __init__(self,
+                 key: str = None,
+                 secret: str = None,
+                 region: str = 'eu-west-2',
+                 service: str = 'api',
+                 version: str = 'latest'
+                 ):
+        self.key = key
+        self.secret = secret
+        self.region = region
+        self.connection = ConnectionUserAndKey(self.key, self.secret)
+        self.connection.region_name = region
+        self.connection.service_name = service
+        self.service_name = service
+        self.version = version
+        self.signer = OSCRequestSignerAlgorithmV4(
+            access_key=self.key,
+            access_secret=self.secret,
+            version=self.version,
+            connection=self.connection
+        )
+        self.NODE_STATE = {
+            'pending': NodeState.PENDING,
+            'running': NodeState.RUNNING,
+            'shutting-down': NodeState.UNKNOWN,
+            'terminated': NodeState.TERMINATED,
+            'stopped': NodeState.STOPPED
+        }
+
+    def list_locations(self, ex_dry_run: bool = False):
+        """
+        Lists available regions details.
+        :param      ex_dry_run: If true, checks whether you have the required
+        permissions to perform the action.
+        :type       ex_dry_run: ``bool``
+        :return: regions details
+        :rtype: ``dict``
+        """
+        action = "ReadRegions"
+        data = json.dumps({"DryRun": ex_dry_run})
+        response = self._call_api(action, data)
+        return self._to_locations(response.json()["Regions"])
+
+    def ex_create_public_ip(self, dry_run: bool = False):
+        """
+        Create a new public ip.
+
+        :param      dry_run: If true, checks whether you have the required
+        permissions to perform the action.
+        :type       dry_run: ``bool``
+
+        :return: the created public ip
+        :rtype: ``dict``
+            """
+        action = "CreatePublicIp"
+        data = json.dumps({"DryRun": dry_run})
+        if self._call_api(action, data).status_code == 200:
+            return True
+        return False
+
+    def ex_delete_public_ip(self,
+                            dry_run: bool = False,
+                            public_ip: str = None,
+                            public_ip_id: str = None):
+        """
+        Delete instances.
```

Comment:
Some comments are out of sync with code, but I will update / fix that when merging the pr :)